Add initial support for prefab struct types

takikawa opened the pull request:
The main thing that is missing from this PR is contract generation support. It turns out that the contract system doesn't provide a suitable contract combinator for prefab structs, so we'll have to make one. I'm planning to follow up with that in a separate PR.

The first few commits and 04ac72e are misc commits that are related (came up while developing) but aren't directly about prefabs.

Other missing features: individually mutable fields, support for auto-values, and typechecking for functions like `make-prefab-struct`.

For the first two, I'm planning to generally improve struct support (not just prefabs) in a separate PR.

takikawa commented on typed-racket-lib/typed-racket/typecheck/tc-structs.rkt:
```diff
@@ -225,9 +228,17 @@
+      (struct-names-type-name (parsed-struct-names parsed-struct))))
   (refine-variance! names stys tvarss))
 
+;; get-struct-name : (U Struct Prefab) -> Symbol
```
Whoops, I forgot that I added this function. I'm planning to delete it since it's unused now.

samth commented on typed-racket-lib/typed-racket/infer/infer-unit.rkt:
```diff
@@ -547,6 +547,12 @@
+          ;; two prefab structs with the same key
+          [((Prefab: k flds) (Prefab: k* flds*))
+           #:when (equal? k k*)
+           ;; FIXME: should account for mutable fields
```
We can't merge this without either a more-conservative approach or a correct approach here.

samth commented on typed-racket-lib/typed-racket/rep/type-rep.rkt:
```diff
@@ -402,9 +402,18 @@
+                  [flds (listof Type/c)])
+  [#:frees (λ (f) (combine-frees (map f flds)))]
+  [#:fold-rhs (*Prefab key (map type-rec-id flds))]
+  [#:key 'prefab])
```
Can prefabs be `procedure?`?

takikawa replied:
No, because prefab struct type declarations cannot use the `#:property` option.

samth commented on typed-racket-lib/typed-racket/typecheck/tc-structs.rkt:
```diff
@@ -272,6 +283,49 @@
   (parsed-struct sty names desc (struct-info-property nm/par) type-only))
 
+;; check and register types for a prefab struct
+(define (tc/prefab vars nm/par fld-names tys
```
This needs a signature. Also, do we want to support `#:type-only?` here?

takikawa replied:
I don't think we need `#:type-only?` here because it's only used for `require/typed`. Since prefab structs can be declared anywhere, there's no need to do a special `require/typed` for them.

samth commented on typed-racket-lib/typed-racket/types/prefab.rkt:
```diff
@@ -0,0 +1,112 @@
+                        (-> prefab-key? prefab-key? any)])
+
+;; Convert a prefab key to its expanded version
+(define (normalize-prefab-key key field-length)
```
I feel like this function must exist somewhere already.

samth commented on typed-racket-lib/typed-racket/types/prefab.rkt:
```diff
@@ -0,0 +1,112 @@
+                 count)))))
+
+;; Convert a prefab key to a shortened version
+(define (abbreviate-prefab-key key)
```
Similarly, this function feels like it belongs somewhere else.

maxsnew replied:
`cddddr` but not `caaddr`?

samth commented on typed-racket-test/unit-tests/typecheck-tests.rkt:
```diff
@@ -93,25 +93,31 @@
 
-  ;; test: syntax? tc-results? [(option/c tc-results?)] -> void?
+  ;; test: syntax? tc-results? [(option/c tc-results?)]
+  ;;       [(listof (list id type))] -> void?
```
Need more explanation of the changes to this file.

takikawa replied:
Do you mean in the commit message, in comments, or elsewhere?

samth replied:
Mostly in the file, but the commit message would be good too.

takikawa commented:
I've now fixed the inference/subtyping for mutable prefab fields. I think I've also addressed the other remaining comments. My plan is to add contract generation separately and push that simultaneously with the `prefab/c` pull request for the core.

samth commented on typed-racket-lib/typed-racket/core.rkt:
```diff
@@ -113,7 +113,8 @@
                               (format "\n[more precisely: ~a]"
-                                      (pretty-format-type (make-Values tcs) #:indent 17))]
+                                      (pretty-format-type (make-Values (map -result tcs))
+                                                          #:indent 17))]
```
Is 17 computed from something?

takikawa replied:
It's computed from the length of the "more precisely" string above in order to get the type to indent to the right position.

samth replied:
Should we have a (computed) constant somewhere for this?

samth commented on typed-racket-lib/typed-racket/infer/infer-unit.rkt:
```diff
@@ -547,6 +547,18 @@
+           #:when (and (prefab-key-subtype? k k*)
+                       (>= (length ss) (length ts)))
+           (% cset-meet*
+              (for/list/fail ([s (in-list (take ss (length ts)))]
```
Why is this `take` call needed? The `for` loop will already stop at the length of `ts`.

takikawa replied:
Good catch, it probably isn't needed.

samth commented on typed-racket-lib/typed-racket/typecheck/tc-literal.rkt:
```diff
@@ -124,8 +125,21 @@
+  (define expected-ts
+    (match (and expected (resolve expected))
+      [(Prefab: _ ts) (in-sequence-forever (in-list ts) #f)]
+      [_ (in-cycle (in-value #f))]))
```
This just throws away the expected type when it's wrong. Can't that miss errors?

takikawa replied:
The result of `tc-literal` is checked against the expected type with `check-below` in `tc-expr/check`, so it should not be possible.

samth commented on typed-racket-lib/typed-racket/typecheck/tc-toplevel.rkt:
```diff
@@ -33,10 +33,17 @@
+           (tc/struct (attribute t.tvars) #'t.nm (syntax->list #'(t.fields ...)) (syntax->list #'(t.types ...))
+                      #:mutable (attribute t.mutable)
+                      #:maker (attribute t.maker)
+                      #:type-only (attribute t.type-only)))]
```
I still feel like these two calls could be combined into one.

takikawa replied:
I'll can try that refactoring and see if it improves the code.

samth commented on typed-racket-lib/typed-racket/types/prefab.rkt:
```diff
@@ -0,0 +1,127 @@
+         racket/list
+         racket/match)
+
+(provide/cond-contract [normalize-prefab-key
```
This code should all go somewhere in core Racket -- see discussion w/ @mflatt elsewhere. But that's fine to do later, once we figure out what we want.

samth commented on typed-racket-lib/typed-racket/types/subtype.rkt:
```diff
@@ -554,6 +554,20 @@
+          (and (prefab-key-subtype? k1 k2)
+               (and (>= (length ss) (length ts))
+                    (for/fold ([A A0])
+                              ([s (in-list (take ss (length ts)))]
```
Again w/ the extra `take`.

takikawa commented:
I removed the redundant `take`s and merged the `tc/struct` and `tc/prefab` functions now.

samth commented:
Looks good.

takikawa commented:
Merged.
